### PR TITLE
breaking: polish @clz() @ctz() @popcount() @bswap() @bitReverse() 

### PR DESCRIPTION
**Pull request by shawnl**
Closes: #2119 

breaking: @bitreverse -> @bitReverse, @bswap -> @bswap

Closes: #2120

Comes with tests

OK. This has changed quite a bit since it was opened. It gained `@fshl` and `@fshr` , and then I added `@memMove` and merged a differn't pull request so I could flesh out `@memMove`. I am working on #767

### CODE REVIEW

File: doc/langref.html.in [hunk 1]
```diff
@@ -6066,7 +6066,7 @@ comptime {
       {#header_close#}
 
       {#header_open|@bSwap#}
-      <pre>{#syntax#}@bSwap(comptime T: type, value: T) T{#endsyntax#}</pre>
+      <pre>{#syntax#}@bSwap(value: var) @typeOf(var){#endsyntax#}</pre>
```
Comment (Hejsil) [inline doc/langref.html.in]: This should probably be `@typeOf(value)`

File: doc/langref.html.in [hunk 2]
```diff
@@ -6075,7 +6075,7 @@ comptime {
       {#header_close#}
 
       {#header_open|@bitReverse#}
-      <pre>{#syntax#}@bitReverse(comptime T: type, integer: T) T{#endsyntax#}</pre>
+      <pre>{#syntax#}@bitReverse(integer: var) @typeOf(var){#endsyntax#}</pre>
```
Comment (Hejsil) [inline doc/langref.html.in]: This should probably be `@typeOf(integer)`

File: doc/langref.html.in [hunk 3]
```diff
@@ -6163,7 +6163,7 @@ comptime {
       {#header_close#}
 
       {#header_open|@clz#}
-      <pre>{#syntax#}@clz(comptime T: type, integer: T) math.Log2Int(@intType(false, @typeOf(T).Int.bits + 1)){#endsyntax#}</pre>
+      <pre>{#syntax#}@clz(integer: var) math.Log2Int(@intType(false, @typeInfo(@typeOf(var)).Int.bits + 1)){#endsyntax#}</pre>
```
Comment (Hejsil) [inline doc/langref.html.in]: `@typeInfo(@typeOf(integer))`

File: doc/langref.html.in [hunk 4]
```diff
@@ -6304,7 +6304,7 @@ test "main" {
       {#header_close#}
 
       {#header_open|@ctz#}
-      <pre>{#syntax#}@ctz(comptime T: type, integer: T) math.Log2Int(@intType(false, @typeOf(T).Int.bits + 1)){#endsyntax#}</pre>
+      <pre>{#syntax#}@ctz(integer: var) math.Log2Int(@intType(false, @typeInfo(@typeOf(var)).Int.bits + 1)){#endsyntax#}</pre>
```
Comment (Hejsil) [inline doc/langref.html.in]: `@typeInfo(@typeOf(integer))`

File: doc/langref.html.in [hunk 5]
```diff
@@ -6862,7 +6862,7 @@ test "call foo" {
       {#header_close#}
 
       {#header_open|@popCount#}
-      <pre>{#syntax#}@popCount(comptime T: type, value: T) math.Log2Int(@intType(false, @typeOf(T).Int.bits + 1)){#endsyntax#}</pre>
+      <pre>{#syntax#}@popCount(integer: var) math.Log2Int(@intType(false, @typeInfo(@typeOf(var)).Int.bits + 1)){#endsyntax#}</pre>
```
Comment (Hejsil) [inline doc/langref.html.in]: `@typeInfo(@typeOf(integer))`

**Comment by andrewrk** [comment]
Why are the new files added?

**Comment by shawnl** [comment]
> Why are the new files added?

Whoops, thanks for spotting that.

**Comment by shawnl** [comment]
I'm now working on exanding the vector features sufficiently to get this algorithm ported to zig https://github.com/cyb70289/utf8 using llvm vectors, rather than assembly intrinsics.

File: std/special/compiler_rt.zig [hunk 6]
```diff
@@ -9,7 +9,7 @@ comptime {
     @export("__getf2", @import("compiler_rt/comparetf2.zig").__getf2, linkage);
 
     if (!is_test) {
-        // only create these aliases when not testing
+        // only create these aliases when not testing -- Yes, I can read the code, but why?
```
Comment (vegecode) [inline std/special/compiler_rt.zig]: Snarky and unhelpful. Maybe ask why and then add constructive comment.

File: std/special/compiler_rt/addXf3.zig [hunk 7]
```diff
@@ -36,19 +36,19 @@ pub extern fn __subtf3(a: f128, b: f128) f128 {
 // TODO: restore inline keyword, see: https://github.com/ziglang/zig/issues/2154
 fn normalize(comptime T: type, significand: *@IntType(false, T.bit_count)) i32 {
     const Z = @IntType(false, T.bit_count);
-    const S = @IntType(false, T.bit_count - @clz(Z(T.bit_count) - 1));
+    const S = @IntType(false, T.bit_count - @clz(@IntType(false, T.bit_count), Z(T.bit_count) - 1));
```
Comment (vegecode) [inline std/special/compiler_rt/addXf3.zig]:
```zig
@clz(Z, Z(T.bit_count) - 1)
```
No need to repeat.

File: std/special/compiler_rt/addXf3.zig [hunk 8]
```diff
@@ -36,19 +36,19 @@ pub extern fn __subtf3(a: f128, b: f128) f128 {
...
     const significandBits = std.math.floatMantissaBits(T);
     const implicitBit = Z(1) << significandBits;
 
-    const shift = @clz(significand.*) - @clz(implicitBit);
+    const shift = @clz(@IntType(false, T.bit_count), significand.*) - @clz(@IntType(false, T.bit_count), implicitBit);
```
Comment (vegecode) [inline std/special/compiler_rt/addXf3.zig]: Use `Z` in both places.

File: std/special/compiler_rt/addXf3.zig [hunk 9]
```diff
@@ -36,19 +36,19 @@ pub extern fn __subtf3(a: f128, b: f128) f128 {
...
     significand.* <<= @intCast(S, shift);
     return 1 - shift;
 }
 
 // TODO: restore inline keyword, see: https://github.com/ziglang/zig/issues/2154
 fn addXf3(comptime T: type, a: T, b: T) T {
     const Z = @IntType(false, T.bit_count);
-    const S = @IntType(false, T.bit_count - @clz(Z(T.bit_count) - 1));
+    const S = @IntType(false, T.bit_count - @clz(@IntType(false, T.bit_count), Z(T.bit_count) - 1));
```
Comment (vegecode) [inline std/special/compiler_rt/addXf3.zig]: Use `Z`

File: std/special/compiler_rt/addXf3.zig [hunk 10]
```diff
@@ -162,7 +162,7 @@ fn addXf3(comptime T: type, a: T, b: T) T {
         // If partial cancellation occured, we need to left-shift the result
         // and adjust the exponent:
         if (aSignificand < implicitBit << 3) {
-            const shift = @intCast(i32, @clz(aSignificand)) - @intCast(i32, @clz(implicitBit << 3));
+            const shift = @intCast(i32, @clz(@IntType(false, T.bit_count), aSignificand)) - @intCast(i32, @clz(@IntType(false, T.bit_count), implicitBit << 3));
```
Comment (vegecode) [inline std/special/compiler_rt/addXf3.zig]: Use `Z` in both places.

File: std/special/compiler_rt/extendXfYf2.zig [hunk 11]
```diff
@@ -69,7 +69,8 @@ inline fn extendXfYf2(comptime dst_t: type, comptime src_t: type, a: src_t) dst_
         // a is denormal.
         // renormalize the significand and clear the leading bit, then insert
         // the correct adjusted exponent in the destination type.
-        const scale: u32 = @clz(aAbs) - @clz(src_rep_t(srcMinNormal));
+        const scale: u32 = @clz(@IntType(false, @typeInfo(src_t).Float.bits), aAbs) -
+                           @clz(@IntType(false, @typeInfo(src_t).Float.bits), src_rep_t(srcMinNormal));
```
Comment (vegecode) [inline std/special/compiler_rt/extendXfYf2.zig]: Use `src_rep_t`

File: std/special/compiler_rt/mulXf3.zig [hunk 12]
```diff
@@ -260,7 +260,7 @@ fn normalize(comptime T: type, significand: *@IntType(false, T.bit_count)) i32 {
     const significandBits = std.math.floatMantissaBits(T);
     const implicitBit = Z(1) << significandBits;
 
-    const shift = @clz(significand.*) - @clz(implicitBit);
+    const shift = @clz(@IntType(false, T.bit_count), significand.*) - @clz(@IntType(false, T.bit_count), implicitBit);
```
Comment (vegecode) [inline std/special/compiler_rt/mulXf3.zig]: Use `Z`

File: test/stage1/behavior/bitreverse.zig [hunk 13]
```diff
@@ -2,80 +2,68 @@ const std = @import("std");
...
     var neg16: i16 = -32694;
-    expect(@bitreverse(i16, -32694) == @bitreverse(i16, neg16));
+    expect(@bitReverse(i16, i16(-32694)) == @bitReverse(i16, neg16));
     var neg24: i24 = -6773785;
-    expect(@bitreverse(i24, -6773785) == @bitreverse(i24, neg24));
+    expect(@bitReverse(i24, i24(-6773785)) == @bitReverse(i24, neg24));
     var neg32: i32 = -16773785;
-    expect(@bitreverse(i32, -16773785) == @bitreverse(i32, neg32));
-    var neg40: i40 = minInt(i40) + 12345;
-    expect(@bitreverse(i40, minInt(i40) + 12345) == @bitreverse(i40, neg40));
```
Comment (vegecode) [inline test/stage1/behavior/bitreverse.zig]: Why remove these?

Comment (shawnl) [reply test/stage1/behavior/bitreverse.zig]: because they are tautologies. They are not testing bitreverse.

**Comment by andrewrk** [comment]
This is too much. I can't keep up with this ever-growing pull request. @shawnl I need you to make one pull request at a time, which only tackles one thing.

File: test/stage1/behavior/bitreverse.zig (hunk 13 above)
Comment (vegecode) [reply test/stage1/behavior/bitreverse.zig]: At the time that code was written, I'm not sure about now, the comment at the top of the block explained correctly what was happening. The LHS values were calculated by comptime code in the Zig compiler while the RHS values are calculated on the LLVM side. These tests were used to make sure the two match.

Comment (andrewrk) [reply test/stage1/behavior/bitreverse.zig]: @vegecode is correct, and it's still true. `@bitreverse(i32, -16773785)` exercises comptime evaluation; `@bitreverse(i32, neg32)` with `neg32` as a `var` exercises runtime evaluation.

**Pull request closed without merging**